### PR TITLE
Let Gradle consume the plugin portal as a regular maven repository

### PR DESCRIPTION
**Pull request by eskatos**
See https://github.com/gradle/kotlin-dsl/issues/469#issuecomment-337493517 for a motivation summary.

### CODE REVIEW

**Comment by eskatos** [comment]
Performance test failure https://builds.gradle.org/viewLog.html?buildId=9047945& happen because these scenarii simulate a slow down on repositories and before the changes in this PR, this slowdown wasn't applied to the Gradle plugin portal, it is now. @oehme and I agreed to rebaseline the tests.

**Comment by oehme** [comment]
Please `@Ignore` the test until the rebaselining is done. That avoids build failures on master which would impact others.

**Comment by eskatos** [comment]
@oehme done ✅

File: subprojects/plugin-use/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequestApplicator.java [hunk 1]
```diff
@@ -65,15 +63,15 @@
 public class DefaultPluginRequestApplicator implements PluginRequestApplicator {
     private final PluginRegistry pluginRegistry;
     private final PluginResolverFactory pluginResolverFactory;
-    private final PluginRepositoryRegistry pluginRepositoryRegistry;
+    private final Factory<List<ArtifactRepository>> pluginRepositoriesFactory;
```
Comment (oehme) [inline subprojects/plugin-use/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequestApplicator.java]: ❌ A factory seems like the wrong abstraction here. The word "Factory" implies I get something new on each call.

File: subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/repository/ResolvingFromMultipleCustomPluginRepositorySpec.groovy [hunk 2]
```diff
@@ -194,8 +195,8 @@ class ResolvingFromMultipleCustomPluginRepositorySpec extends AbstractDependency
 
         then:
         failure.assertThatDescription(containsNormalizedString("""
-- ${repoType}(${repoA.uri}) (Could not resolve plugin artifact 'org.gradle.hello-world:org.gradle.hello-world.gradle.plugin:0.2')
-- ${repoType}(${repoB.uri}) (Could not resolve plugin artifact 'org.gradle.hello-world:org.gradle.hello-world.gradle.plugin:0.2')"""
+- ${repoType} (Could not resolve plugin artifact 'org.gradle.hello-world:org.gradle.hello-world.gradle.plugin:0.2')
```
[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/repository/ResolvingFromMultipleCustomPluginRepositorySpec.groovy]: Not all repositories have a URL. Some encompasses multiple locations. etc...

We could either:
1. let `ArtifactRepository implements Describable` and use that, but coming up with the right strings is not trivial ;
2. or reuse the dependency resolution exception and its nice multi-line error message.

I'll give `#2` a shot

[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/repository/ResolvingFromMultipleCustomPluginRepositorySpec.groovy]: Normal dependency resolution failures looks like this: https://scans.gradle.com/s/k73f552wewbiw/console-log

After thinking a bit more, point `#1` above would be a breaking change, and `#2` is not that easy to get right but feels like the right way to go.
I tend towards keeping my changes as is and implement #2 as a follow up.
WWYT?

[the remark this answers is not in the document]
Comment (oehme) [reply subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/repository/ResolvingFromMultipleCustomPluginRepositorySpec.groovy]: Sounds good

**Comment by eskatos** [comment]
CI [Branch Accept](https://builds.gradle.org/viewLog.html?buildId=9128000&) ✅ 

@oehme, could you have another look?

**Comment by oehme** [comment]
Looks good to me 👍

**Comment by eskatos** [comment]
Follow up issue: https://github.com/gradle/gradle/issues/3416

Merging

**Pull request merged**